### PR TITLE
[ci] Install Kythe. Extract and upload Kzips.

### PR DESCRIPTION
**Pull request by dmcardle**
This PR is necessary, but not sufficient, to enable semantic indexing on https://cs.opensource.google/opentitan. Once we get this working, I can update a Google-internal config to start using the uploaded Kzips.

#13935

### CODE REVIEW

**Comment by milesdai** [comment]
> Is it even possible for CI to pass on this kind of CI-modifying PR?

Yes, changes made to CI by a PR will be incorporated in that PR's CI run.

The error happening here in this PR comes from this line: `long="verilator-version:,openocd-version:,verible-version:,rust-version:"` in `ci/install-package-dependencies.sh`. Just need to add kythe-version to the option-parsing.

**Comment by dmcardle** [comment]
Solid point, @GregAC. Chatted with @milesdai and we'll add a GitHub action for this.

**Comment by dmcardle** [comment]
@milesdai PTAL! This is a little more complicated than the hello-world we talked about, but it stops short of uploading to a GCS bucket.

I set this up to run periodically via "cron" rather than every push to master. Once it's chugging along, it might make sense to increase the frequency.

**Review by milesdai — COMMENTED** [review]
Look like a great start! Just left some initial thoughts on this

File: .github/workflows/kythe.yml
Comment (GregAC) [reply .github/workflows/kythe.yml:18]: I don't think that's necessary here. For the main CI we are testing OpenTitan itself functions correctly with the documentation minimum environment (which is based on Ubuntu 18.04). This is just publishing OT code to another index, it's not something that's really part of supported OT functionality so it's reasonable to just do whatever needed for it to work and we can ignore the standard OT environment setup.

File: .github/workflows/kythe.yml
Comment (dmcardle) [reply .github/workflows/kythe.yml:11]: I don't care strongly in the longer term, but I want the task to repeat slowly at first so I can monitor it.

I just blindly followed the Kythe team's recommendation to make a periodic, 6-hour task (would link, but it's not a public doc). I'm actually not sure that cs.opensource.google would actually re-index at the rate that we merge PRs, so a bunch of kzip uploads might be wasted.

File: .github/workflows/kythe.yml
Comment (dmcardle) [reply .github/workflows/kythe.yml:18]: One side benefit of using the latest release is that it might give us a heads up on distro changes that would break our build. The downside is a lack of uniformity and that we might run into unique bugs that will never happen on the real builders. Happy to embrace the chaos if Miles is.

**Review by dmcardle — COMMENTED** [review]
@milesdai Are you aware of a way to test GitHub actions/workflows locally? Do we have to merge it to see what happens?

File: .github/workflows/kythe.yml
Comment (dmcardle) [inline .github/workflows/kythe.yml:31]: I suppose we should check the SHA256 sum here.

Comment (dmcardle) [reply .github/workflows/kythe.yml:31]: Done.

**Comment by milesdai** [comment]
> @milesdai Are you aware of a way to test GitHub actions/workflows locally? Do we have to merge it to see what happens?

I don't think there's a way to test locally on your machine, but in my past experience, I think if you configure your action to run using a PR trigger, you can have it run on just your PR to test it out. If you have it set as a scheduled task (like it is currently), then I think you'll have to wait until it's merged.

**Comment by dmcardle** [comment]
> > @milesdai Are you aware of a way to test GitHub actions/workflows locally? Do we have to merge it to see what happens?
> 
> I don't think there's a way to test locally on your machine, but in my past experience, I think if you configure your action to run using a PR trigger, you can have it run on just your PR to test it out. If you have it set as a scheduled task (like it is currently), then I think you'll have to wait until it's merged.

Ah, thanks! Just tacked on a commit to change the trigger to when the PR is opened or edited (I hope that's only code changes, not comments and the like).

**Comment by dmcardle** [comment]
@milesdai, PTAL!

FYI, this new action is currently failing because we need to configure a GitHub secret with the GCP credentials, but I need an admin for that.

File: .github/workflows/kythe.yml
Comment (milesdai) [inline .github/workflows/kythe.yml:56]: nit: would it be better to check the hash first before extracting?

File: .github/workflows/kythe.yml
Comment (milesdai) [inline .github/workflows/kythe.yml:46]: Just to stay consistent with azure-pipelines.yml, can we remove the newlines between each command?

**Review by milesdai — APPROVED** [review]
Small nits, but otherwise LGTM! Happy to take a look after getting the GCP creds.

File: .github/workflows/kythe.yml
Comment (dmcardle) [reply .github/workflows/kythe.yml:56]: Yes, absolutely.

File: .github/workflows/kythe.yml
Comment (dmcardle) [reply .github/workflows/kythe.yml:46]: Done.

**Comment by GregAC** [comment]
Ok I think we can get the secret via an environment. I've created one called 'Kythe' that has the appropriate secret. I think you'll need to make a modification to your workflow to run in the environment: https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment#using-an-environment

When we get that to work remove the environment from the workflow and then we merge. Then when running against master it should have access to the secret.

Could really use a button to say 'temporarily give this PR access to all repository secrets' to test this kind of stuff!

**Comment by GregAC** [comment]
Oh and for extra paranoia only myself and @dmcardle can approve actions in PR to run under the Kythe environment.

**Comment by dmcardle** [comment]
Neat, that environment trick seemed to work — I got an email asking me to approve the action for this PR.

**Comment by dmcardle** [comment]
Hmm, setting `environment: Kythe` didn't help. The latest run[^1] failed with this message:

> google-github-actions/auth failed with: retry function failed with 0 attempts: the GitHub Action workflow must specify exactly one of "workload_identity_provider" or "credentials_json"! If you are specifying input values via GitHub secrets, ensure the secret is being injected into the environment. By default, secrets are not passed to workflows triggered from forks, including Dependabot.

I'm providing only `credentials_json`, so I suspect the issue is that GitHub is still not injecting secrets into this PR because it came from my fork.

@milesdai, WDYT about merging to see if the upload step works? I wouldn't normally recommend testing in prod, but I don't think there's much danger. It wouldn't block other people's PRs with spurious failures because it only runs on master. (First, I'd need to resolve the open TODOs in the patch.)

[^1]: https://github.com/lowRISC/opentitan/actions/runs/2805202474

File: .github/workflows/kythe.yml
Comment (GregAC) [inline .github/workflows/kythe.yml:37]:
```suggestion
    environment: Kythe
    env:
      GCP_CREDENTIALS_KYTHE_UPLOADER: ${{ secrets.GCP_CREDENTIALS_KYTHE_UPLOADER }}
```

File: .github/workflows/kythe.yml
Comment (GregAC) [inline .github/workflows/kythe.yml:78]:
```suggestion
          credentials_json: '${{ env.GCP_CREDENTIALS_KYTHE_UPLOADER }}'
```

**Review by GregAC — COMMENTED** [review]
I think you might need the suggested additions to get access to the secret (see https://docs.github.com/en/actions/security-guides/encrypted-secrets#using-encrypted-secrets-in-a-workflow) however even with that with a bit of searching around I don't think there's any way to get a secret into a PR from a fork (The 'environment' referred to in the error message is the shell environment, not the github actions environment I think....).

I could create a kythe_test branch you could push to then we try a PR from that?

**Comment by dmcardle** [comment]
> I think you might need the suggested additions to get access to the secret (see https://docs.github.com/en/actions/security-guides/encrypted-secrets#using-encrypted-secrets-in-a-workflow) however even with that with a bit of searching around I don't think there's any way to get a secret into a PR from a fork (The 'environment' referred to in the error message is the shell environment, not the github actions environment I think....).

I'm following the [example from `google-github-actions/auth`](https://github.com/google-github-actions/auth#authenticating-via-service-account-key-json-1) and it just specifies the `credentials_json` input without touching the shell environment. I'd also like to avoid setting an environment variable so we don't accidentally log our secret in the future — it would only take an accidental `env` shell command in this action to leak the secret.

> I could create a kythe_test branch you could push to then we try a PR from that?

Let's try that! Per the note at the top of the doc you linked, I think there may no way to access in this fork-initiated PR.

**Comment by GregAC** [comment]
Ok kythe_test branch created, not entirely sure if you can push to it though, give it a go and see what happens.

**Comment by dmcardle** [comment]
> Ok kythe_test branch created, not entirely sure if you can push to it though, give it a go and see what happens.

I did some git gymnastics locally, but no dice. I set up a copy of my local branch named `kythe_test`, set the upstream to `origin/kythe_test` (where origin is `git@github.com:lowRISC/opentitan.git`), and tried to push.

```console
% git push origin
ERROR: Permission to lowRISC/opentitan.git denied to dmcardle.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

Is it possible to give me permission to push to that branch?

**Comment by dmcardle** [comment]
@GregAC Don't bother monkeying with branch permissions — it occurred to me that I can test this in a PR directed at my own fork.

I'll need to mirror the secret configuration on my fork, but the GitHub action shouldn't have a problem accessing the secret. This should be enough to demonstrate that the action successfully uploads to the GCS bucket, after which I'll modify *this* PR so it only runs on the master branch.

File: .github/workflows/kythe.yml
Comment (moidx) [inline .github/workflows/kythe.yml:20]: Are you planning to remove these before submitting? I assume that you are still working on getting this to turn green in your presubmit?

**Pull request closed without merging**
